### PR TITLE
Added grid_sample backward batch rule

### PR DESCRIPTION
**Pull request by vfdev-5**
Description:
- Added grid_sample backward batch rule: CPU and CUDA
- Updated tests
- Updated codegen/codegen_outofplacebatching.py to be able to parse methods with `std::array` in signature
- Regenerated functorch/csrc/OutOfPlacePlumbing.cpp

Notes:
I had to expand on dim 0 in most of the cases and could not use
tricks like in forward pass when batch dim is merged either with channel or H_out
due to wrong grid grads in these cases

Related to #240

### CODE REVIEW

File: codegen/codegen_outofplacebatching.py [hunk 1]
```diff
@@ -155,7 +155,7 @@ def parse_return(return_t):
     return tuple([x.strip() for x in m.group(1).split(',')])
 
 def parse_args(args_t):
-    args = args_t.split(',')
+    args = args_t.split(', ')
```
Comment (vfdev-5) [inline codegen/codegen_outofplacebatching.py:158]: There is an assumption made that args are separated with comma-space. Such that I do not reinvent how to parse args in general case with any kind of arg types.

Comment (zou3519) [reply codegen/codegen_outofplacebatching.py:158]: This is fine, can you add a comment here regarding the assumption?

File: functorch/csrc/BatchRulesModules.cpp [hunk 2]
```diff
@@ -400,7 +643,12 @@ TORCH_LIBRARY_IMPL(aten, FT_BATCHED_KEY, m) {
   EXISTING_BDIM(im2col_backward);
 
   VMAP_SUPPORT("grid_sampler_2d", GRID_SAMPLE_BATCH_RULE(grid_sampler));
+  VMAP_SUPPORT("grid_sampler_2d_backward", GRID_SAMPLE_BW_BATCH_RULE(grid_sampler_2d_backward));
+
   VMAP_SUPPORT("grid_sampler_3d", GRID_SAMPLE_BATCH_RULE(grid_sampler));
+  VMAP_SUPPORT("grid_sampler_3d_backward", GRID_SAMPLE_BW_BATCH_RULE(grid_sampler_3d_backward));
+  VMAP_SUPPORT("cudnn_grid_sampler_backward", CUDNN_GRID_SAMPLE_BW_BATCH_RULE(cudnn_grid_sampler_backward));
```
Comment (zou3519) [inline functorch/csrc/BatchRulesModules.cpp:650]: Do both these get exercised in the tests?

Comment (vfdev-5) [reply functorch/csrc/BatchRulesModules.cpp:650]: Yes, `cudnn_grid_sampler_backward` is tested by test/test_ops.py::TestOperatorsCUDA::test_vmapvjp_has_batch_rule_nn_functional_grid_sample_cuda_float32 
and `grid_sampler_3d_backward` is tested by test/test_ops.py::TestOperatorsCPU::test_vmapvjp_has_batch_rule_nn_functional_grid_sample_cpu_float32

File: functorch/csrc/BatchRulesModules.cpp [hunk 3]
```diff
@@ -255,6 +255,202 @@ grid_sample_batch_rule(const Tensor& input, optional<int64_t> input_bdim, const
...
+  auto new_grad_output = grad_output;
+  auto new_input = input;
+  auto new_grid = grid;
+
+  optional<int64_t> grad_input_out_bdim = nullopt;
+  optional<int64_t> grad_grid_out_bdim = nullopt;
+  int64_t bdim_size = 0;
+
+  if (grad_output_bdim) {
+
```
[the remark this answers is not in the document]
Comment (vfdev-5) [reply functorch/csrc/BatchRulesModules.cpp:281]: Yes, that's true that we can simplify everything into a similar code. Thanks

**Review by zou3519 — APPROVED** [review]
thanks!

**Comment by samdow** [comment]
Test is failing because we're getting an unexpected success from `test_vmapvjp_nn_functional_pad_circular_cpu_float32`

I'm also not sure if this is related but just to flag if it is

**Comment by vfdev-5** [comment]
There a cuda test failing at build due to pytorch 1.7.1 installation by conda (somehow), https://app.circleci.com/pipelines/github/pytorch/functorch/798/workflows/e842f064-bc17-4925-b330-88f0aa0c94a1/jobs/4776?invite=true#step-103-333

**Comment by samdow** [comment]
> There is a cuda test failing at build due to pytorch 1.7.1 installation by conda (somehow), https://app.circleci.com/pipelines/github/pytorch/functorch/798/workflows/e842f064-bc17-4925-b330-88f0aa0c94a1/jobs/4776?invite=true#step-103-333

Just reran the tests (sorry that it reran all of them...I mean to only do that one). It looks like that just hit a weird race condition or something funky happened with the checkout since the root errors were a bunch of "file not found"s. I think the CUDA test failures should be similar across the 3 versions after this

**Comment by vfdev-5** [comment]
I synced the PR to the `main` branch as `main` is passing now and let's see if something still breaks

**Pull request merged**